_posts/2025-05-28-#32051.md

```diff
@@ -7,25 +7,102 @@ authors: [jonatack]
+or stalling logic, `ThreadOpenAddedConnections` attempts to immediately
+reconnect it -- unless "onetry" was passed to the addnode RPC -- up to the limit
+of 8 addnode connections. This limit is separate from the regular peer
+connection limits.
+
+`ThreadOpenAddedConnections` will continue to attempt reconnection of the
+disconnected addnode peer until it succeeds.
+
+When these disconnection/reconnection cycles happen frequently with
+addnode peers, it is likely network, resource and time intensive.
+This is particularly true for I2P peers, as these involve destroying and
+rebuilding 2 tunnels for each peer connection. It seems worth avoiding this if
+it is straightforward to do so.
+
+Automatic (non-addnode) peers are also disconnected by the same logic, but they
+are a different category and case (non-protected peers, no immediate
+connection/reconnection) that would require monitoring over time to adjust the
+timeouts accordingly. Martin Zumsande was looking into optimizing this (see
+https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2025-01-22#1083993): *"The
+challenge is to distinguish this situation from making things worse for
+fast/reliable connections that just have some slow peers which should be
+disconnected."*
+
+The goal of this pull request is thus to avoid unnecessary frequent
+disconnections and immediate reconnections of addnode peers, both during IBD and
+afterwards.
+
+### Approach
+
+1. The first commit, "p2p: protect addnode peers during IBD", provides addnode
+   peers the max `BLOCK_STALLING_TIMEOUT_MAX` value of 64 seconds for the IBD
+   stalling logic ("Peer is stalling block download") in `src/net_processing.cpp`.
 
+2. The second commit, "p2p: don't disconnect addnode peers for block download
+   timeout", proposes to protect addnode peers from disconnection. [Review
+   feedback](https://github.com/bitcoin/bitcoin/pull/32051#pullrequestreview-2691505746)
+   suggested that we also clear their block requests, so that these blocks can
+   be requested from other peers.
+
+3. The third commit, "p2p: don't disconnect addnode peers for slow
+   initial-headers-sync", proposes the same protection for addnode peers that we
+   currently already provide to peers with `NetPermissionFlags::NoBan`
+   permission.
+
+4. The fourth commit, "rpc, doc: update addnode documentation", updates the RPC
+   addnode help documentation.
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+2. What is an addnode peer? How can you specify them to your node?
+
+3. Do you provide addnode peers to the node(s) that you run? Why or why not?
+   What kind of peers do you choose?
+
+4. What is a fundamental protection that addnode peers can provide to your node?
+
+5. It appears the issue mostly, but not exclusively, arises during IBD.
+   Which code is responsible for that discrepancy?
```

Comment:
"Peer is stalling block download, disconnecting" only happens during IBD, I believe.

Reply:
"Timeout downloading block...disconnecting" -> after IBD

Reply:
I'm not familiar with this code, but am I misunderstand this docstring then? https://github.com/bitcoin/bitcoin/pull/32051/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R5792-R5794

Reply:
Alright, I'll revert this question, sorry for the noise.